packages/react-router/src/scroll-restoration.tsx

```diff
@@ -1,241 +1,315 @@
-import * as React from 'react'
 import { functionalUpdate } from '@tanstack/router-core'
 import { useRouter } from './useRouter'
+import { ScriptOnce } from './ScriptOnce'
+import type { AnyRouter } from './router'
 import type { NonNullableUpdater, ParsedLocation } from '@tanstack/router-core'
 
-const useLayoutEffect =
-  typeof window !== 'undefined' ? React.useLayoutEffect : React.useEffect
+export type ScrollRestorationEntry = { scrollX: number; scrollY: number }
 
-const windowKey = 'window'
-const delimiter = '___'
+export type ScrollRestorationByElement = Record<string, ScrollRestorationEntry>
 
-let weakScrolledElements = new WeakSet<any>()
+export type ScrollRestorationByKey = Record<string, ScrollRestorationByElement>
 
-type CacheValue = Record<string, { scrollX: number; scrollY: number }>
-type CacheState = {
-  cached: CacheValue
-  next: CacheValue
+export type ScrollRestorationCache = {
+  state: ScrollRestorationByKey
+  set: (updater: NonNullableUpdater<ScrollRestorationByKey>) => void
 }
-
-type Cache = {
-  state: CacheState
-  set: (updater: NonNullableUpdater<CacheState>) => void
+export type ScrollRestorationOptions = {
+  getKey?: (location: ParsedLocation) => string
+  scrollBehavior?: ScrollToOptions['behavior']
 }
 
+export const storageKey = 'tsr-scroll-restoration-v3'
```

Comment:
```suggestion
// TODO: Bump `v1` to `v2` when performing the library major bump.
export const storageKey = 'tsr-scroll-restoration-v1-3'
```

And then in v2, change it to `tsr-scroll-restoration-v2-1`.